### PR TITLE
multi-stage build

### PR DESCRIPTION
**Pull request by andrevtg**
This is a proposal to use a multi-stage build Dockerfile. A previous builder image containing the proper build tools results in a faster (and maybe cleaner) build.

Codefresh.io already deals with multi-stage builds, unlike Docker hub (at this date).

### CODE REVIEW

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: 72b8e9783514b6c97279dc16d551fc3ae6256fc8
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Comment by karalabe** [comment]
I've also wrote up a commit already that does exactly this. Still, we're holding it off a bit until the multi-stage stuff gets mainstream enough. There's no point to roll out until most people upgrade, otherwise it will just be a mess :)

File: Dockerfile [hunk 1]
```diff
@@ -1,12 +1,14 @@
-FROM alpine:3.5
+FROM vertigo/go-builder as builder
```
Comment (karalabe) [inline Dockerfile]: We won't accept a non official build base image btw. Our solution was to use the official go builder image + install anything missing on top of it.

File: Dockerfile [hunk 2]
```diff
@@ -1,12 +1,14 @@
...
 
 ADD . /go-ethereum
 RUN \
-  apk add --update git go make gcc musl-dev linux-headers && \
   (cd go-ethereum && make geth)                           && \
   cp go-ethereum/build/bin/geth /geth                     && \
-  apk del git go make gcc musl-dev linux-headers          && \
-  rm -rf /go-ethereum && rm -rf /var/cache/apk/*
+  rm -rf /go-ethereum
```
Comment (karalabe) [inline Dockerfile]: There's no reason to delete anything from the builder image, it's going to get discarded anyway.

File: Dockerfile [hunk 3]
```diff
@@ -1,12 +1,14 @@
-FROM alpine:3.5
+FROM vertigo/go-builder as builder
 
 ADD . /go-ethereum
 RUN \
-  apk add --update git go make gcc musl-dev linux-headers && \
   (cd go-ethereum && make geth)                           && \
   cp go-ethereum/build/bin/geth /geth                     && \
```
Comment (karalabe) [inline Dockerfile]: There's no reason to copy geth to a different location, you can copy it from it's build path directly into the final image.

File: Dockerfile (hunk 1 above)
Comment (andrevtg) [inline Dockerfile]: I am aware of that. That was just a proposal, pointing to the multi-staging direction. All the rest was just copy/pasted from the original Dockerfile, I overlooked the other details.

Glad to know this has been taken care of already.

**Comment by andrevtg** [comment]
Closing.

**Pull request closed without merging**